File: backend/internal/musicbrainz/handler.go
package musicbrainz

import (
	"encoding/json"
	"net/http"

	"github.com/go-chi/chi/v5"
)

type Handler struct {
	client *Client
}

func NewHandler(client *Client) *Handler {
	return &Handler{client: client}
}

func (h *Handler) SearchAlbums(w http.ResponseWriter, r *http.Request) {
	title := r.URL.Query().Get("title")
	artist := r.URL.Query().Get("artist")
	releases, err := h.client.SearchAlbums(title, artist, 20)
	if err != nil {
		w.WriteHeader(503)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	err = json.NewEncoder(w).Encode(releases)
	if err != nil {
		w.WriteHeader(503)
	}
}

func (h *Handler) RegisterRoutes(r chi.Router) {
	r.Get("/albums", h.SearchAlbums)
}

File: backend/src/musicbrainz/handler.ts
import { Hono } from 'hono'
import { MusicBrainzClient, RateLimiter } from './client.js'

export function musicBrainzRoutes(client = new MusicBrainzClient(new RateLimiter(1000))) {
  const routes = new Hono()

  routes.get('/albums', async (c) => {
    const title = c.req.query('title') ?? ''
    const artist = c.req.query('artist') ?? ''
    try {
      const releases = await client.searchAlbums(title, artist, 20)
      return c.json(releases)
    } catch {
      return c.body(null, 503)
    }
  })

  return routes
}
